Add serverless io manager appropriate usage for baa (#22544)
dagster-io/dagster: docs/content/dagster-plus/deployment/serverless.mdx

Unlike Hybrid, Serverless Deployments on Dagster+ require direct access to your data, secrets and source code.

- Secrets and source code are built into the image directly. Images are stored in a per-customer container registry with restricted access.
- User code is securely sandboxed using modern container sandboxing techniques.
- All production access is governed by industry-standard best practices which are regularly audited.

### IO Management on Serverless

> <Warning>
> If you are a Serverless customer who is working on personally identifiable information (PII), private health information (PHI), has signed a business association agreement (BAA), or are otherwise working with data subject to GDPR or other such regulations, you must not use the default IO manager.
> </Warning> 

Reviewer: ```suggestion
<Note>
The default I/O manager cannot be used if you are a Serverless user who:
<ul>
<li>Works with personally identifiable information (PII)</li>
<li>Works with private health information (PHI)</li>
<li>Has signed a business association agreement (BAA), or</li>
<li>Are otherwise working with data subject to GDPR or other such regulations</li>
</ul>
</Note> 
```

A few suggestions here:

I know `Warning` is techincally the correct admonition here, but our current styling has the callout blend into the page:

![Screenshot 2024-06-14 at 11 57 50 AM](https://github.com/dagster-io/dagster/assets/16615212/af7ea6d8-59e5-4881-b26f-2c3321f17fe9)

Given how important this is, I think using `Note` to make this stand out is a good idea.

I also suggest we lead with "You can't use the default I/O manager" so the thing we're warning people about comes first. I also put the criteria into a list to make it skimmable. (I know the HTML is ugly, but soon we won't have to do this in callouts!)

Thoughts?

> In Serverless, code that uses the default [IO Manager](/concepts/io-management/io-managers#built-in-io-managers) is automatically adjusted to save data in Dagster Plus managed storage. This automatic change is useful because the default file system in Serverless is ephemeral which means the default IO manager would not work as expected. However, this automatic change means potentially sensitive data is being stored, not just processed or orchestrated, by Dagster Plus. To avoid this behavior:
> - Use an IO manager that stores data in your infrastructure
> - Write code that does not use an IO manager

Reviewer: ```suggestion
In Serverless, code that uses the default [I/O manager](/concepts/io-management/io-managers#built-in-io-managers) is automatically adjusted to save data in Dagster+ managed storage. This automatic change is useful because the default file system in Serverless is ephemeral, which means the default I/O manager wouldn't work as expected. However, this automatic change means potentially sensitive data is being stored, not just processed or orchestrated, by Dagster+. 

To avoid this behavior, you can:

- Use an I/O manager that stores data in your infrastructure
- Write code that doesn't use an I/O manager
```

- `Dagster Plus` should be `Dagster+`
- We write `IO` as `I/O`, no capitalization needed in `manager`
- Added a comma or two to break up some longer sentences
